pjmedia/src/pjmedia-videodev/android_dev.c

```diff
@@ -1109,6 +1143,162 @@ static pj_status_t and_stream_destroy(pjmedia_vid_dev_stream *s)
     return PJ_SUCCESS;
 }
 
+#if USE_CAMERA2
+
+static void JNICALL OnGetFrame2(JNIEnv *env, jobject obj,
+				jlong user_data,
+				jobject plane0, jint rowStride0, jint pixStride0,
+				jobject plane1, jint rowStride1, jint pixStride1,
+				jobject plane2, jint rowStride2, jint pixStride2)
+{
+    and_stream *strm = (and_stream*)(intptr_t)user_data;
+    pjmedia_frame f;
+    pj_uint8_t *p0, *p1, *p2;
+    jlong p0_len, p1_len, p2_len;
+    pj_uint8_t *Y, *U, *V;
+    pj_status_t status;
+    void *frame_buf, *data_buf;
+    
+    strm->frame_ts.u64 += strm->ts_inc;
+    if (!strm->vid_cb.capture_cb)
+	return;
+
+    if (strm->thread_initialized == 0 || !pj_thread_is_registered()) {
+	pj_status_t status;
+	pj_bzero(strm->thread_desc, sizeof(pj_thread_desc));
+	status = pj_thread_register("and_cam", strm->thread_desc,
+				    &strm->thread);
+	if (status != PJ_SUCCESS)
+	    return;
+	strm->thread_initialized = 1;
+	PJ_LOG(5,(THIS_FILE, "Android camera thread registered"));
+    }
+
+    p0 = (pj_uint8_t*)(*env)->GetDirectBufferAddress(env, plane0);
+    p1 = (pj_uint8_t*)(*env)->GetDirectBufferAddress(env, plane1);
+    p2 = (pj_uint8_t*)(*env)->GetDirectBufferAddress(env, plane2);
+    p0_len = (*env)->GetDirectBufferCapacity(env, plane0);
+    p1_len = (*env)->GetDirectBufferCapacity(env, plane1);
+    p2_len = (*env)->GetDirectBufferCapacity(env, plane2);
+    
+    /* Assuming the buffers are originally a large contigue buffer */
+    pj_assert(p1 < p0+strm->vafp.framebytes && p2 < p0+strm->vafp.framebytes);
+
+    f.type = PJMEDIA_FRAME_TYPE_VIDEO;
+    f.size = strm->vafp.framebytes;
+    f.timestamp.u64 = strm->frame_ts.u64;
+    f.buf = data_buf = p0;
+
+    Y = (pj_uint8_t*)f.buf;
+    U = Y + strm->vafp.plane_bytes[0];
+    V = U + strm->vafp.plane_bytes[1];
+
+    /* Check if we need conversion here, this is the tricky part of camera2.
+     * When we request I420, the returned buffer may not be actually I420,
+     * for example it is NV21 in Samsung S10. The camera2 'cheats' us via
```

Comment:
Maybe it's better to move device-specific testing to the issue desc/comment instead. One reason is that there will be timestamp that can be easily tracked in the issue page.

Reply:
The code comment is about a sample case & details of the buffer trick by camera2 which is closely related to the conversion checks done in the following code, instead of a specific device test result. Anyway, the device may exhibit different behavior in the future which can be confusing, so I just stripped it out.